### PR TITLE
Sprint22 matwerk #248

### PR DESCRIPTION
**Pull request by mkuehbach**

### CODE REVIEW

File: contributed_definitions/nyaml/NXapm_compositionspace_config.yaml [hunk 1]
```diff
@@ -1,25 +1,12 @@
...
-    Number of voxel of discretized domain for analyzed part of the dataset.
-  d: |
-    The dimensionality of the grid.
-  c: |
-    The cardinality or total number of cells/grid points.
-  n_clst_dict: |
-    Number of terms in the composition clustering dictionary
-  n_spat_dict: |
-    Number of terms in the position clustering dictionary
+  * `A. Saxena et al. <https://www.github.com/eisenforschung/CompositionSpacetool.git>`_
```
Comment (lukaspie) [inline contributed_definitions/nyaml/NXapm_compositionspace_config.yaml:5]: Link seems to be broken, I guess it's https://github.com/eisenforschung/CompositionSpace?

File: contributed_definitions/nyaml/NXapm_compositionspace_config.yaml [hunk 2]
```diff
@@ -34,61 +21,97 @@ NXapm_compositionspace_config(NXobject):
...
-      edge_length(NX_FLOAT):
+          doc: |
+            Name of the (parent) node directly below which the ranging definitions for (molecular) ions are stored.
+      voxelization(NXprocess):
         doc: |
-          Edge length of the cubic voxels that is used for discretizing the point cloud in the voxelization step.
-        unit: NX_LENGTH
-      bics_clusters(NX_UINT):
+          Step during which the point cloud is discretized to compute element-specific composition fields.
+          Iontypes are atomically decomposed to correctly account for the multiplicity of each element that
```
Comment (lukaspie) [inline contributed_definitions/nyaml/NXapm_compositionspace_config.yaml:55]: This was a bit hard to understand, maybe rephrase / explain a bit better.

File: contributed_definitions/nyaml/NXapm_compositionspace_config.yaml [hunk 3]
```diff
@@ -34,61 +21,97 @@ NXapm_compositionspace_config(NXobject):
...
+            Edge length of cubic voxels for 3D grid that is used for discretizing the point cloud.
+          unit: NX_LENGTH
+      autophase(NXprocess):
         doc: |
-          The maximum number of components configured for the segmentation of the voxel
-          set into a mixture of voxels with that many different chemical compositions.
-        unit: NX_UNITLESS
-      ml_models(NXobject):
+          Optional step during which the subsequent segmentation step is prepared to improve the segmentation.
+        use(NX_BOOLEAN):
```
Comment (lukaspie) [inline contributed_definitions/nyaml/NXapm_compositionspace_config.yaml:64]: maybe call it `used`. Alternative, make it the NXprocess optional/recommended and if it is there, it is assumed it was used.

File: contributed_definitions/nyaml/NXapm_compositionspace_config.yaml [hunk 4]
```diff
@@ -34,61 +21,97 @@ NXapm_compositionspace_config(NXobject):
...
+          doc: |
+            Edge length of cubic voxels for 3D grid that is used for discretizing the point cloud.
+          unit: NX_LENGTH
+      autophase(NXprocess):
         doc: |
-          The maximum number of components configured for the segmentation of the voxel
-          set into a mixture of voxels with that many different chemical compositions.
-        unit: NX_UNITLESS
-      ml_models(NXobject):
+          Optional step during which the subsequent segmentation step is prepared to improve the segmentation.
```
Comment (lukaspie) [inline contributed_definitions/nyaml/NXapm_compositionspace_config.yaml:63]: If this is an optional step, why not make it optional in the appdef?

File: contributed_definitions/nyaml/NXapm_compositionspace_config.yaml [hunk 5]
```diff
@@ -34,61 +21,97 @@ NXapm_compositionspace_config(NXobject):
...
+        use(NX_BOOLEAN):
+          doc: |
+            Was the automated phase assignment used?
+            This algorithm achieves a supervised reduction of the dimensionality of the chemical space
```
Comment (lukaspie) [inline contributed_definitions/nyaml/NXapm_compositionspace_config.yaml:67]: This should probably go to the main doc of autophase

File: contributed_definitions/nyaml/NXapm_compositionspace_results.yaml [hunk 6]
```diff
@@ -1,19 +1,21 @@
...
 doc: |
-  Results of a run with Alaukik Saxena's composition space tool.
+  Application definition for results of the CompositionSpace tool used in atom probe.
   
-  This is an initial draft application definition for the common NFDI-MatWerk,
-  FAIRmat infrastructure use case IUC09 how to improve the organization and
-  results storage of the composition space tool and make these data at the
-  same time directly understandable for research data management systems
-  like NOMAD Oasis.
+  * `A. Saxena et al. <https://www.github.com/eisenforschung/CompositionSpacetool.git>`_
```
Comment (lukaspie) [inline contributed_definitions/nyaml/NXapm_compositionspace_results.yaml:5]: again link broken

File: contributed_definitions/nyaml/NXapm_compositionspace_results.yaml [hunk 7]
```diff
@@ -1,19 +1,21 @@
...
+  
+  This is a draft application definition for the common NFDI-MatWerk/FAIRmat infrastructure
+  use case IUC09 that explores how to improve the organization and results storage of the
+  CompositionSpace tool.
 symbols:
   doc: |
     The symbols used in the schema to specify e.g. dimensions of arrays.
   d: |
```
Comment (lukaspie) [inline contributed_definitions/nyaml/NXapm_compositionspace_results.yaml:13]: I would suggest to use better symbols (like `grid_dim`, `n_voxels`).

**Review by lukaspie — CHANGES_REQUESTED** [review]
Cannot comment much about the contents, so mostly syntax and best practices.

Two questions:
1) There are a lot of required elements, are they always there? Especially those that are stated to be optional in the docs.
2) Is it possible to merge the two application definitions described here? I.e., have the configs and results for each process together? I don't know too much about the workflow, so maybe that's nonsense.

File: contributed_definitions/nyaml/NXapm_compositionspace_config.yaml (hunk 1 above)
Comment (mkuehbach) [reply contributed_definitions/nyaml/NXapm_compositionspace_config.yaml:5]: Thank you that was a mistake indeed

File: contributed_definitions/nyaml/NXapm_compositionspace_config.yaml [hunk 8]
```diff
@@ -34,61 +21,97 @@ NXapm_compositionspace_config(NXobject):
...
+          to use for this analysis.
         type(NX_CHAR):
         path(NX_CHAR):
         checksum(NX_CHAR):
         algorithm(NX_CHAR):
         position(NX_CHAR):
-        mass_to_charge(NX_CHAR):
+          doc: |
+            Name of the node which resolves the reconstructed ion position values to use for this analysis.
+        # mass_to_charge(NX_CHAR):
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply contributed_definitions/nyaml/NXapm_compositionspace_config.yaml:38]: Made optional, and made docstring clear

File: contributed_definitions/nyaml/NXapm_compositionspace_config.yaml [hunk 9]
```diff
@@ -34,61 +21,97 @@ NXapm_compositionspace_config(NXobject):
...
+        doc: |
+          Specification of the tomographic reconstruction used for this analysis.
+          Typically, reconstructions in the field of atom probe tomography are communicated
+          via files which store at least reconstructed ion positions and mass-to-charge-state-ratio
+          values. Container files like HDF5 though can store multiple reconstructions.
+          Therefore, the position and mass_to_charge concepts point to specific instances
+          to use for this analysis.
         type(NX_CHAR):
         path(NX_CHAR):
         checksum(NX_CHAR):
```
[the remark this answers is not in the document]
Comment (mkuehbach) [reply contributed_definitions/nyaml/NXapm_compositionspace_config.yaml:33]: Constraint set adjusted, you got the idea

File: contributed_definitions/nyaml/NXapm_compositionspace_config.yaml (hunk 4 above)
Comment (mkuehbach) [reply contributed_definitions/nyaml/NXapm_compositionspace_config.yaml:63]: Fixed

File: contributed_definitions/nyaml/NXapm_compositionspace_config.yaml (hunk 5 above)
Comment (mkuehbach) [reply contributed_definitions/nyaml/NXapm_compositionspace_config.yaml:67]: Docstring reorganized as suggested

File: contributed_definitions/nyaml/NXapm_compositionspace_results.yaml (hunk 6 above)
Comment (mkuehbach) [reply contributed_definitions/nyaml/NXapm_compositionspace_results.yaml:5]: Indeed, thank you

File: contributed_definitions/nyaml/NXapm_compositionspace_results.yaml (hunk 7 above)
Comment (mkuehbach) [reply contributed_definitions/nyaml/NXapm_compositionspace_results.yaml:13]: Accepted and implemented

**Comment by mkuehbach** [comment]
1. There are a lot of required elements, are they always there? Especially those that are stated to be optional in the docs.
> I have adjusted this slightly, if an analysis step has been executed what was computed is documented

>     2. Is it possible to merge the two application definitions described here? I.e., have the configs and results for each process together? I don't know too much about the workflow, so maybe that's nonsense.

> Technically yes but I do not want this the idea of these workflows is you have a well-defined configuration + well-defined input you pass this to a blackbox (which does some processing based on the cfg and input) that returns an artifact bundling results together, the results are not so large in data volume nor out-of-core computing or parallel I/O is required therefore all can go together in one output file, HDF5 internally stores time data thus blending input and output together would require one to either recreate the config file if you run on another day another run with say a different identifier.
For this reason I decided to split apart always in my tools config from results.

**Comment by lukaspie** [comment]
> > 2. Is it possible to merge the two application definitions described here? I.e., have the configs and results for each process together? I don't know too much about the workflow, so maybe that's nonsense.
> 
> Technically yes but I do not want this the idea of these workflows is you have a well-defined configuration + well-defined input you pass this to a blackbox (which does some processing based on the cfg and input) that returns an artifact bundling results together, the results are not so large in data volume nor out-of-core computing or parallel I/O is required therefore all can go together in one output file, HDF5 internally stores time data thus blending input and output together would require one to either recreate the config file if you run on another day another run with say a different identifier. For this reason I decided to split apart always in my tools config from results.

Ok, makes sense, I'll approve then.

**Review by lukaspie — APPROVED** [review]
LGTM
